bucket_test.go

```diff
@@ -58,3 +59,50 @@ func TestBucket_Take(t *testing.T) {
 		}
 	}
 }
+
+func TestBucket_Merge(t *testing.T) {
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+	buckets := make([]Bucket, 100)
+	for i := range buckets {
+		buckets[i] = Bucket{
+			Added: rng.Float64(), // The P of the PN counter "tokens".
+			Taken: rng.Float64(), // The N of the PN counter "tokens".
+			Last:  rng.Int63(),   // A separate "last" timestamp G-Counter.
+		}
+	}
+
+	// Compute the result of a merged bucket with sequential operations.
+	var sequential Bucket
+	for _, bucket := range buckets {
+		sequential.Merge(&bucket)
+	}
+
+	// Compute multiple random sequences of merge operations and compare with
+	// the sequential result. With this, we test that the Merge operation
+	// makes the Bucket a CRDT by holding the following properties independently
+	// of merge order:
+	//
+	// - Associativity (a+(b+c)=(a+b)+c)
+	// - Commutativity (a+b=b+a)
+	// - Idempotence (a+a=a)
+	//
+	for i := 0; i < 10000; i++ { // Test 10000 random permutations of Merge order.
+		rng.Shuffle(len(buckets), func(i, j int) {
+			buckets[i], buckets[j] = buckets[j], buckets[i]
+		})
+
+		random := buckets[rng.Int()%len(buckets)]
+		for _, bucket := range buckets {
+			random.Merge(&bucket)
```

Comment:
To test idempotence, it might be nice to add e.g.

```go
if (rng.Int() % 100) > 66 {
    random.Merge(&bucket) // again
}
```

Reply:
Since we start with a random Bucket and merge it with itself already, is this necessary?

Reply:
Ah, so you do exercise idempotence with n=1. I don't know! Perhaps not, but perhaps it's worth exaggerating e.g. network retransmissions, for comedic effect.